File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/scoreboard/SurfScoreboardImpl.java
package dev.slne.surf.surfapi.bukkit.server.scoreboard;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import dev.slne.surf.surfapi.bukkit.api.SurfBukkitApi;
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfScoreboard;
import java.util.List;
import net.kyori.adventure.text.Component;
import net.megavex.scoreboardlibrary.api.sidebar.Sidebar;
import net.megavex.scoreboardlibrary.api.sidebar.component.ComponentSidebarLayout;
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.FramedSidebarAnimation;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Nullable;

@ApiStatus.Internal
public class SurfScoreboardImpl implements SurfScoreboard {

  protected final Component title;
  protected final int maxLines;
  protected final SidebarComponent sidebarComponent;
  protected final List<SidebarAnimation<Component>> animations;
  protected @Nullable Sidebar scoreboard;
  protected @Nullable ComponentSidebarLayout sidebarLayout;
  protected boolean enabled = false;

  public SurfScoreboardImpl(Component title, int maxLines, SidebarComponent sidebarComponent,
      List<SidebarAnimation<Component>> animations) {
    this.title = title;
    this.maxLines = maxLines;
    this.sidebarComponent = sidebarComponent;
    this.animations = animations;
  }

  @Override
  public void addViewer(Player viewer) {
    checkNotNull(viewer, "viewer");
    checkState(enabled, "Scoreboard is not enabled. Did you forget to call enable()?");

    assert scoreboard != null : "scoreboard is null";
    scoreboard.addPlayer(viewer);
  }

  @Override
  public void removeViewer(Player viewer) {
    checkNotNull(viewer, "viewer");
    checkState(enabled, "Scoreboard is not enabled. Did you forget to call enable()?");

    assert scoreboard != null : "scoreboard is null";
    scoreboard.removePlayer(viewer);
  }

  @Override
  public void enable() {
    checkState(!enabled, "Scoreboard is already enabled");

    scoreboard = SurfBukkitApi.get().getScoreboardLibrary().createSidebar(maxLines);
    sidebarLayout = new ComponentSidebarLayout(SidebarComponent.staticLine(title),
        sidebarComponent);

    sidebarLayout.apply(scoreboard);

    enabled = true;
  }

  @Override
  public void disable() {
    checkState(enabled, "Scoreboard is not enabled. Did you forget to call enable()?");

    assert scoreboard != null : "scoreboard is null";

    scoreboard.close();
    animations.forEach(frameSwitcher -> {
      if (frameSwitcher instanceof FramedSidebarAnimation<Component> animation) {
        animation.switchFrame(0);
      }
    });

    scoreboard = null;
    sidebarLayout = null;
    enabled = false;
  }

  @Override
  public void update() {
    checkState(enabled, "Scoreboard is not enabled. Did you forget to call enable()?");

    assert scoreboard != null : "scoreboard is null";
    assert sidebarLayout != null : "sidebarLayout is null";

    animations.forEach(SidebarAnimation::nextFrame);
    sidebarLayout.apply(scoreboard);
  }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/scoreboard/SurfScoreboardImpl.kt
package dev.slne.surf.surfapi.bukkit.server.scoreboard

import dev.slne.surf.surfapi.bukkit.api.scoreboard.ObsoleteScoreboardApi
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfScoreboard
import dev.slne.surf.surfapi.bukkit.api.surfBukkitApi
import net.kyori.adventure.text.Component
import net.megavex.scoreboardlibrary.api.sidebar.Sidebar
import net.megavex.scoreboardlibrary.api.sidebar.component.ComponentSidebarLayout
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.FramedSidebarAnimation
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation
import org.bukkit.entity.Player

@ObsoleteScoreboardApi
open class SurfScoreboardImpl(
    protected val title: Component,
    protected val maxLines: Int,
    protected val sidebarComponent: SidebarComponent,
    protected val animations: List<SidebarAnimation<Component>>
) : SurfScoreboard {

    protected var scoreboard: Sidebar? = null
    protected var sidebarLayout: ComponentSidebarLayout? = null
    protected var enabled: Boolean = false

    override fun addViewer(viewer: Player) {
        check(enabled) { "Scoreboard is not enabled. Did you forget to call enable()?" }

        scoreboard!!.addPlayer(viewer)
    }

    override fun removeViewer(viewer: Player) {
        check(enabled) { "Scoreboard is not enabled. Did you forget to call enable()?" }

        scoreboard!!.removePlayer(viewer)
    }

    override fun enable() {
        check(!enabled) { "Scoreboard is already enabled" }

        scoreboard = surfBukkitApi.scoreboardLibrary().createSidebar(maxLines)
        sidebarLayout = ComponentSidebarLayout(
            SidebarComponent.staticLine(title),
            sidebarComponent
        ).also { it.apply(scoreboard!!) }

        enabled = true
    }

    override fun disable() {
        check(enabled) { "Scoreboard is not enabled. Did you forget to call enable()?" }

        scoreboard!!.close()
        animations.forEach { (it as? FramedSidebarAnimation<Component>)?.switchFrame(0) }

        this.scoreboard = null
        sidebarLayout = null
        enabled = false
    }

    override fun update() {
        check(enabled) { "Scoreboard is not enabled. Did you forget to call enable()?" }

        animations.forEach { it.nextFrame() }
        sidebarLayout!!.apply(scoreboard!!)
    }
}
